viewer/latexworkshop.ts

```diff
@@ -304,46 +304,100 @@ class LateXWorkshopPdfViewer implements ILatexWorkshopPdfViewer {
         this.sendCurrentStateToPanelManager()
     }
 
-    private forwardSynctex(position: { page: number, x: number, y: number, indicator: boolean }) {
+    private forwardSynctex(data: SynctexData | SynctexRangeData[]) {
         if (!this.synctexEnabled) {
             this.addLogMessage('SyncTeX temporarily disabled.')
             return
         }
-        // use the offsetTop of the actual page, much more accurate than multiplying the offsetHeight of the first page
-        // https://github.com/James-Yu/LaTeX-Workshop/pull/417
+        
         const container = document.getElementById('viewerContainer') as HTMLElement
-        const pos = PDFViewerApplication.pdfViewer._pages[position.page - 1].viewport.convertToViewportPoint(position.x, position.y)
-        const page = document.getElementsByClassName('page')[position.page - 1] as HTMLElement
-        const maxScrollX = window.innerWidth * 0.9
-        const minScrollX = window.innerWidth * 0.1
-        let scrollX = page.offsetLeft + pos[0]
-        scrollX = Math.min(scrollX, maxScrollX)
-        scrollX = Math.max(scrollX, minScrollX)
-        const scrollY = page.offsetTop + page.offsetHeight - pos[1]
-
-        // set positions before and after SyncTeX to viewerHistory
-        this.viewerHistory.set(container.scrollTop)
-        if (PDFViewerApplication.pdfViewer.scrollMode === 1) {
-            // horizontal scrolling
-            container.scrollLeft = page.offsetLeft
+
+        if (Array.isArray(data)){
+            const indicatorTemplate = document.getElementById('synctex-indicator') as HTMLElement
+            const indicatorParent = indicatorTemplate.parentNode as HTMLElement
+
+            for (const record of data) {
+                const pos_left_top = PDFViewerApplication.pdfViewer._pages[record.page - 1].viewport.convertToViewportPoint(record.h, record.v - record.H)
+                const pos_right_down = PDFViewerApplication.pdfViewer._pages[record.page - 1].viewport.convertToViewportPoint(record.h + record.W, record.v)
+    
+                const width_px = pos_right_down[0] - pos_left_top[0]
```

Comment:
This new implementation and below until end of function is a bit long. I'd advice add key comments to if and for conditions at least, to improve readability.